File: problems/atcoder/abc128_b/main.cpp
/**
 * Author by Rainboy blog: https://rainboylv.com github: https://github.com/rainboylvx
 * rbook: -> https://rbook.roj.ac.cn  https://rbook2.roj.ac.cn
 * rainboy的学习导航网站: https://idx.roj.ac.cn
 * create_at: 2026-07-10 21:42
 * update_at: 2026-07-10 21:42
 */
#include <bits/stdc++.h>
using namespace std;

struct Rest {
    string city;
    int score, id;
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n; cin >> n;
    vector<Rest> a(n);
    for (int i = 0; i < n; i++) {
        cin >> a[i].city >> a[i].score;
        a[i].id = i + 1;
    }

    sort(a.begin(), a.end(), [](const Rest &x, const Rest &y) {
        if (x.city != y.city) return x.city < y.city;
        return x.score > y.score;
    });

    for (auto &r : a) cout << r.id << '\n';

    return 0;
}

File: problems/atcoder/abc128_b/main.hs
{-
 Author by Rainboy blog: https://rainboylv.com github: https://github.com/rainboylvx
 rbook: -> https://rbook.roj.ac.cn  https://rbook2.roj.ac.cn
 rainboy的学习导航网站: https://idx.roj.ac.cn
 create_at: 2026-07-10 22:00
 update_at: 2026-07-10 22:00
-}
import Data.List (sortOn)
import Data.Ord (Down(..))

type Restaurant = (String, Int, Int)  -- (city, score, index)

parseLine :: Int -> String -> Restaurant
parseLine i line =
    let [city, s] = words line
    in (city, read s, i)

main = do
    input <- getContents
    let (_:lns) = lines input
        restaurants = zipWith parseLine [1..] lns
        sorted = sortOn (\(city, score, _) -> (city, Down score)) restaurants
    mapM_ (print . (\(_,_,i) -> i)) sorted
